com.unity.render-pipelines.universal/Editor/ShaderGraph/Includes/ShaderPassDecal.hlsl

```diff
@@ -0,0 +1,306 @@
+{
+    surfaceData.albedo = decalSurfaceData.baseColor.rgb;
+    surfaceData.metallic = saturate(decalSurfaceData.metallic);
+    surfaceData.specular = 0;
+    surfaceData.smoothness = saturate(decalSurfaceData.smoothness);
+    surfaceData.occlusion = decalSurfaceData.occlusion;
+    surfaceData.emission = decalSurfaceData.emissive;
+    surfaceData.alpha = saturate(decalSurfaceData.baseColor.w);
+    surfaceData.clearCoatMask = 0;
+    surfaceData.clearCoatSmoothness = 1;
+}
+
+PackedVaryings Vert(Attributes inputMesh)
+{
+    Varyings output = (Varyings)0;
+#ifdef DECAL_MESH
+    if (_DecalMeshBiasType == DECALMESHDEPTHBIASTYPE_VIEW_BIAS) // TODO: Check performance of branch
+    {
+        float3 viewDirectionOS = GetObjectSpaceNormalizeViewDir(inputMesh.positionOS);
+        inputMesh.positionOS += viewDirectionOS * (_DecalMeshViewBias);
+    }
+    output = BuildVaryings(inputMesh);
+    if (_DecalMeshBiasType == DECALMESHDEPTHBIASTYPE_DEPTH_BIAS) // TODO: Check performance of branch
+    {
+        MeshDecalsPositionZBias(output);
+    }
+#else
+    output = BuildVaryings(inputMesh);
+#endif
+
+#ifdef VARYINGS_NEED_LIGHTMAP_UV
+    OUTPUT_LIGHTMAP_UV(inputMesh.uv1, unity_LightmapST, output.lightmapUV);
+#endif
+
+#ifdef VARYINGS_NEED_SH
+    half3 sh = 0;
+    OUTPUT_SH(half3(output.normalWS), sh);
+    output.sh = float3(sh);
+#endif
+
+    PackedVaryings packedOutput = (PackedVaryings)0;
+    packedOutput = PackVaryings(output);
+
+    return packedOutput;
+}
+
+void Frag(PackedVaryings packedInput,
+#if defined(DECAL_DBUFFER)
+    OUTPUT_DBUFFER(outDBuffer)
+#elif defined(DECAL_SCREEN_SPACE)
+    out half4 outColor : SV_Target0
+#elif defined(DECAL_GBUFFER)
+    out FragmentOutput fragmentOutput
+#elif defined(DECAL_FORWARD_EMISSIVE)
+    out half4 outEmissive : SV_Target0
+#elif defined(SCENEPICKINGPASS)
+    out float4 outColor : SV_Target0
+#else
+#error SHADERPASS_is_not_correctly_define
+#endif
+)
+{
+#ifdef SCENEPICKINGPASS
+    outColor = _SelectionID;
+#else
+    UNITY_SETUP_STEREO_EYE_INDEX_POST_VERTEX(packedInput);
+    UNITY_SETUP_INSTANCE_ID(packedInput);
+    Varyings input = UnpackVaryings(packedInput);
+
+    half angleFadeFactor = 1.0;
+
+#if defined(DECAL_PROJECTOR)
+#if UNITY_REVERSED_Z
+    float depth = LoadSceneDepth(input.positionCS.xy);
+#else
+    // Adjust z to match NDC for OpenGL
+    float depth = lerp(UNITY_NEAR_CLIP_VALUE, 1, LoadSceneDepth(input.positionCS.xy));
+#endif
+#endif
+
+#if defined(DECAL_RECONSTRUCT_NORMAL)
+    #if defined(_DECAL_NORMAL_BLEND_HIGH)
+        half3 normalWS = half3(ReconstructNormalTap9(input.positionCS.xy));
+    #elif defined(_DECAL_NORMAL_BLEND_MEDIUM)
+        half3 normalWS = half3(ReconstructNormalTap5(input.positionCS.xy));
+    #else
+        half3 normalWS = half3(ReconstructNormalDerivative(input.positionCS.xy));
+    #endif
+#elif defined(DECAL_LOAD_NORMAL)
+    half3 normalWS = half3(LoadSceneNormals(input.positionCS.xy));
```

Comment:
if none of DECAL_RECONSTRUCT_NORMAL and DECAL_LOAD_NORMAL is define then 
normalWS isn't define.

Reply:
That should not be reachable, in case it does i would like error